### ISSUE TITLE
Not expected to run ExecutionPlan more than once

### ISSUE THREAD

**Issue by killme2008** [GreptimeTeam/greptimedb#790]
### What type of bug is this?

Unexpected error

### What subsystems are affected?

Datanode

### What happened?

Create a table and insert some data:

```sql
CREATE TABLE integers(i INTEGER, j BIGINT TIME INDEX);

INSERT INTO integers VALUES (1, 1), (2, 2), (3, 3), (NULL, 4);
```

Query the table as below:

```sql
SELECT i1.i,i2.i FROM integers i1 LEFT OUTER JOIN integers i2 ON 1=1 WHERE i1.i>2 ORDER BY 2;
```

Raise an error:

```text
Failed to execute, error: Datanode { code: 1003, msg: "Failed to collect RecordBatches, 
source: Failed to poll stream, source: External error: Arrow error: External error:
 External error: Not expected to run ExecutionPlan more than once" }
```

### What operating system did you use?

Mac osx

### How can we reproduce the bug?

```bash
Described above.
```

**Comment by waynexia** [comment]
Cannot reproduce this in DF's unit test. Maybe this is our problem?

**Comment by killme2008** [comment]
Yes, it's at https://github.com/GreptimeTeam/greptimedb/blob/develop/src/table/src/table/scan.rs#L78

**Comment by Fedomn** [comment]
I'd like to try this. 
After some investigation, I found the optimizer added a `RepartitionExec` operator in physical plan tree, and this operator result in parallel scan on `SimpleTableScan`. Therefore, `SimpleTableScan` will throw above error when invoked secondly.

But I found `SimpleTableScan.execute` method ignore `partition` parameter, seems the underlying table scan function not support the parameter.

So, How about the workaround that returns `EmptyRecordBatchStream` directly when `SimpleTableScan` operator be executed secondly?

**Comment by killme2008** [comment]
@Fedomn Thanks for your investgation. Multi regions are not supported in the table engine right now, @v0y4g3r is working on it #653 , so the `partition` parameter is ignored in `SimpleTableScan.execute`.

I think we can fix this later, after supporting multi regions.

**Comment by evenyag** [comment]
We have set the output partitioning of the plan to 1, so `RepartitionExec` should work fine
https://github.com/GreptimeTeam/greptimedb/blob/26848f9f5c082b4533ac2463e7a51b777355e3c8/src/table/src/table/scan.rs#L60-L62

The problem seems to relate to `JOIN`, or selecting the table twice in the same query. I guess the cause is that datafusion shares and reuses the same `Arc<dyn ExecutionPlan>` multiple times, which breaks our assumption. We might need to refactor our `ExecutionPlan` implementation to allow it to execute repeatedly.

**Comment by Fedomn** [comment]
> We have set the output partitioning of the plan to 1, so `RepartitionExec` should work fine

But above mentioned `RepartitionExec` is added by datafusion optimizer, and the trigger point is `WHERE i1.i>2`.

https://github.com/apache/arrow-datafusion/blob/b686b68baeee964a5b797caf49484c01f0ae5b54/datafusion/core/src/physical_optimizer/repartition.rs#L197-L199

Could use `explain verbose` to see `RepartitionExec` how to added.


> I guess the cause is that datafusion shares and reuses the same `Arc<dyn ExecutionPlan>` multiple times, which breaks our assumption

Yep, the operator will execute the child multi-times in parallel

https://github.com/apache/arrow-datafusion/blob/b686b68baeee964a5b797caf49484c01f0ae5b54/datafusion/core/src/physical_plan/repartition.rs#L444-L446

> We might need to refactor our `ExecutionPlan` implementation to allow it to execute repeatedly.

Agree, maybe datafusion CsvExec could refer to.

https://github.com/apache/arrow-datafusion/blob/b686b68baeee964a5b797caf49484c01f0ae5b54/datafusion/core/src/physical_plan/file_format/csv.rs#L151-L153

**Comment by evenyag** [comment]
@Fedomn Thanks, so there are two different issues we might need to solve:
- The first one is that we need to implement `ExecutionPlan` correctly and defer the actual scan to the execution of the stream.
- Another issue is that we don't support the `partition` param, or in other words, we only support one output partition.

> Yep, the operator will execute the child multi-times in parallel

The `RepartitionExec` uses `input`'s partition count to execute it so I think the `partition` param should always be `0`. The parallelism is restricted by the input partition count, which should be 1 for our `SimpleTableScan` plan.
- https://github.com/apache/arrow-datafusion/blob/b686b68baeee964a5b797caf49484c01f0ae5b54/datafusion/core/src/physical_plan/repartition.rs#L340
- https://github.com/apache/arrow-datafusion/blob/b686b68baeee964a5b797caf49484c01f0ae5b54/datafusion/core/src/physical_plan/mod.rs#L496

Note that the partitioning in `explain verbose` is the output partitioning of the `RepartitionExec`, not its input's. We could see it is also added when we execute `SELECT i1.i FROM integers i1 WHERE i1.i>2;`, which works fine.
```
ProjectionExec: expr=[i@0 as i]
  CoalesceBatchesExec: target_batch_size=4096
    FilterExec: i@0 > 2
      RepartitionExec: partitioning=RoundRobinBatch(10)
        ExecutionPlan(PlaceHolder)
```

Thus the second one isn't a critical problem, though we are also working on the multi-region support.

For the first problem, we can refer to the [datafusion::physical_plan::file_format::FileStream](https://docs.rs/datafusion/latest/datafusion/physical_plan/file_format/struct.FileStream.html) to fix it. But I'm afraid that I don't have time to work on this.

**Comment by evenyag** [comment]
@Fedomn Thanks, so there are two different issues we might need to solve:
- The first one is that we need to implement `ExecutionPlan` correctly and defer the actual scan to the execution of the stream.
- Another issue is that we don't support the `partition` param, or in other words, we only support one output partition.

> Yep, the operator will execute the child multi-times in parallel

The `RepartitionExec` uses `input`'s partition count to execute it so I think the `partition` param should always be `0`. The parallelism is restricted by the input partition count, which should be 1 for our `SimpleTableScan` plan.
- https://github.com/apache/arrow-datafusion/blob/b686b68baeee964a5b797caf49484c01f0ae5b54/datafusion/core/src/physical_plan/repartition.rs#L340
- https://github.com/apache/arrow-datafusion/blob/b686b68baeee964a5b797caf49484c01f0ae5b54/datafusion/core/src/physical_plan/mod.rs#L496

Note that the partitioning in `explain verbose` is the output partitioning of the `RepartitionExec`, not its input's. We could see it is also added when we execute `SELECT i1.i FROM integers i1 WHERE i1.i>2;`, which works fine.
```
ProjectionExec: expr=[i@0 as i]
  CoalesceBatchesExec: target_batch_size=4096
    FilterExec: i@0 > 2
      RepartitionExec: partitioning=RoundRobinBatch(10)
        ExecutionPlan(PlaceHolder)
```

Thus the second one isn't a critical problem, though we are also working on the multi-region support.

For the first problem, we can refer to the [datafusion::physical_plan::file_format::FileStream](https://docs.rs/datafusion/latest/datafusion/physical_plan/file_format/struct.FileStream.html) to fix it. But I'm afraid that I don't have time to work on this.
